src/algebra/cubic_discriminant.lean

```diff
@@ -0,0 +1,316 @@
+/-
+Copyright (c) 2022 David Kurniadi Angdinata. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Kurniadi Angdinata
+-/
+
+import field_theory.splitting_field
+
+/-!
+# Cubics and discriminants
+
+This file defines cubic polynomials over a semiring and their discriminants over a splitting field.
+
+## Main definitions
+
+* `cubic`: the structure representing a cubic polynomial.
+* `disc`: the discriminant of a cubic polynomial.
+
+## Main statements
+
+* `disc_ne_zero_iff_roots_nodup`: the cubic discriminant is not equal to zero if and only if
+  the cubic has no duplicate roots.
+
+## References
+
+* https://en.wikipedia.org/wiki/Cubic_equation
+* https://en.wikipedia.org/wiki/Discriminant
+
+## Tags
+
+cubic, discriminant, polynomial, root
+-/
+
+noncomputable theory
+
+/-- The structure representing a cubic polynomial. -/
+@[ext] structure cubic (R : Type*) := (a b c d : R)
+
+namespace cubic
+
+open cubic polynomial
+
+variables {R S F K : Type*}
+
+instance [inhabited R] : inhabited (cubic R) := ⟨⟨default R, default R, default R, default R⟩⟩
+
+instance [has_zero R] : has_zero (cubic R) := ⟨⟨0, 0, 0, 0⟩⟩
+
+section basic
+
+variables {P : cubic R} [semiring R]
+
+/-- Convert a cubic polynomial to a polynomial. -/
+def to_poly (P : cubic R) : polynomial R := C P.a * X ^ 3 + C P.b * X ^ 2 + C P.c * X + C P.d
+
+/-! ### Coefficients -/
+
+section coeff
+
+private lemma coeffs :
+  P.to_poly.coeff 3 = P.a ∧ P.to_poly.coeff 2 = P.b ∧ P.to_poly.coeff 1 = P.c
+    ∧ P.to_poly.coeff 0 = P.d :=
+begin
+  rw [to_poly],
+  simp only [coeff_add, coeff_C, coeff_C_mul_X, coeff_C_mul_X_pow],
+  change P.a + 0 + 0 + 0 = _ ∧ 0 + P.b + 0 + 0 = _ ∧ 0 + 0 + P.c + 0 = _ ∧ 0 + 0 + 0 + P.d = _,
+  simp only [zero_add, add_zero],
+  exact ⟨rfl, rfl, rfl, rfl⟩
+end
```

Comment:
```suggestion
by simp [to_poly, coeff_C, coeff_C_mul_X, coeff_C_mul_X_pow]
```

Reply:
I'm still not sold with this, but I guess it makes it cleaner, so I have changed it.